src/v/cloud_storage/tests/cloud_storage_e2e_fuzz_test.cc

```diff
@@ -0,0 +1,336 @@
+
+    ss::sstring to_string() const {
+        return ssx::sformat(
+          "{{ {}, {}, {} }}", op_names[static_cast<int>(type)], arg1, arg2);
+    }
+};
+
+const std::vector<std::pair<op_type, int>> op_weights{
+  {op_type::produce, 5},
+  {op_type::local_roll, 2},
+  {op_type::local_gc, 2},
+  {op_type::local_gc_one, 3},
+  {op_type::local_compact, 3},
+  {op_type::upload_segments, 3},
+  {op_type::upload_manifest, 5},
+  {op_type::switch_leader, 2},
+};
+
+test_op random_operation() {
+    static const int weight_total = [&] {
+        int total = 0;
+        for (const auto& [_, weight] : op_weights) {
+            total += weight;
+        }
+        return total;
+    }();
+    auto r = random_generators::get_int(0, weight_total);
+    int sum = 0;
+    op_type type{op_type::max};
+    for (const auto& [op, weight] : op_weights) {
+        sum += weight;
+        if (sum > r) {
+            type = op;
+            break;
+        }
+    }
+    auto arg1 = random_generators::get_int(0, 100);
+    auto arg2 = random_generators::get_int(0, 100);
+    return {type, arg1, arg2};
+}
+
+std::vector<test_op> generate_workload(size_t count) {
+    std::vector<test_op> ret;
+    ret.reserve(count);
+    bool replicated_in_term = true;
+    const int num_nodes = 2;
+    const int max_records = 10;
+    const int max_cardinality = 5;
+    int cur_leader_idx = 0;
+    int num_records = 0;
+    int num_records_uploaded = 0;
+    int num_records_in_manifest = 0;
+
+    // The number of expected local segments per node. We shouldn't try to GC
+    // more than this number of segments.
+    std::vector<int> num_local_segs_per_node(num_nodes);
+
+    // Whether the active segment on each node has any offsets, in which case
+    // we are eligible to roll the segment.
+    std::vector<bool> offsets_in_active_seg_per_node(num_nodes);
+    auto set_offsets_in_active_seg_all_nodes = [&] {
+        for (int i = 0; i < num_nodes; i++) {
+            offsets_in_active_seg_per_node[i] = true;
+        }
+    };
+    // Do some best effort cleanup of randomly generated operations by
+    // enforcing some rough invariants for scheduled operations (e.g. only
+    // roll the local log when there's data to roll).
+    while (ret.size() < count) {
+        auto op = random_operation();
+        int node_idx = op.arg1 % num_nodes;
+        auto& num_local_segs = num_local_segs_per_node[node_idx];
+        auto active_segment_non_empty
+          = offsets_in_active_seg_per_node[node_idx];
+        switch (op.type) {
+        case op_type::produce:
+            replicated_in_term = true;
+            set_offsets_in_active_seg_all_nodes();
+            mod_at_least_one(&op.arg1, max_records);
+            mod_at_least_one(&op.arg2, max_cardinality);
+            num_records += op.arg1;
+            break;
+
+        case op_type::local_roll:
+            // Only roll if we've added to the log since last roll.
+            if (active_segment_non_empty) {
```

Comment:
shouldn't this be `!active_segment_non_empty`?

Reply:
Nice catch, done